### PR TITLE
Support pre_exec in Proxy

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What problem does this PR solve?

Issue Number: ref #5170

Problem Summary:

1. Support pre_exec in Proxy
2. Filter CompactLog if a prelim flush fails
3. Always filter ComputeHash and VerifyHash

We also use a isolated Proxy branch here for easier rollback when failure.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp [hunk 1]
```diff
@@ -54,6 +55,66 @@ class RegionKVStoreTest : public ::testing::Test
     static void testRaftMergeRollback(KVStore & kvs, TMTContext & tmt);
 };
 
+void RegionKVStoreTest::testNewProxy()
+{
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp:59]:
```suggestion
TEST_F(RegionKVStoreTest, NewProxy)
try
{
```

File: dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp [hunk 2]
```diff
@@ -54,6 +55,66 @@ class RegionKVStoreTest : public ::testing::Test
...
+        request.mutable_compact_log();
+        request.set_cmd_type(::raft_cmdpb::AdminCmdType::CompactLog);
+        // CompactLog always returns true now, even if we can't do a flush.
+        // We use a tryFlushData to pre-filter.
+        ASSERT_EQ(kvs.handleAdminRaftCmd(std::move(request), std::move(response), 1, 5, 1, ctx.getTMTContext()), EngineStoreApplyRes::Persist);
+
+        // Filter
+        ASSERT_EQ(kvs.tryFlushRegionData(1, false, ctx.getTMTContext()), false);
+    }
+}
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp:116]:
```suggestion
}
CATCH
```

File: dbms/src/Storages/Transaction/KVStore.cpp [hunk 3]
```diff
@@ -410,19 +410,7 @@ EngineStoreApplyRes KVStore::handleUselessAdminRaftCmd(
...
-            return false;
-        }
-        else
-        {
-            return canFlushRegionDataImpl(curr_region_ptr, true, /* try_until_succeed */ false, tmt, region_task_lock);
-        }
-    };
-
-    if (check_sync_log())
+    if (cmd_type == raft_cmdpb::AdminCmdType::CompactLog)
```
Comment (lidezhu) [inline dbms/src/Storages/Transaction/KVStore.cpp:413]: Please add some comment and links to related proxy code.

**Comment by JaySon-Huang** [comment]
Could we return an enum value {`Done`/`Skip`} instead of boolean for `AdminObserver:: pre_exec_admin`/`QueryObserver:: pre_exec_query`?

https://github.com/pingcap/tidb-engine-ext/compare/fbd8e6f209952d5bb79897d8bce17230acf23495...e83a37d2d8d8ae1778fe279c5f06a851f8c9e56a#diff-6342e87aaabe7cc62e18d32b39683f4db7c3566756769c6d2c296188d40bd805R90-R93

**Comment by CalvinNeo** [comment]
> Could we return an enum value {`Done`/`Skip`} instead of boolean for `AdminObserver:: pre_exec_admin`/`QueryObserver:: pre_exec_query`?
> 
> [pingcap/tidb-engine-ext@fbd8e6f...e83a37ddiff-6342e87aaabe7cc62e18d32b39683f4db7c3566756769c6d2c296188d40bd805R90-R93](https://github.com/pingcap/tidb-engine-ext/compare/fbd8e6f209952d5bb79897d8bce17230acf23495...e83a37d2d8d8ae1778fe279c5f06a851f8c9e56adiff-6342e87aaabe7cc62e18d32b39683f4db7c3566756769c6d2c296188d40bd805R90-R93)

LGTM. However, we'd better use this solution for now.

**Review by JaySon-Huang — APPROVED** [review]
lgtm

**Review by hehechen — APPROVED** [review]

**Pull request merged**